doobie/src/main/scala/tamer/db/package.scala

```diff
@@ -0,0 +1,137 @@
+      setup: Setup[K, V, S]
+  )(state: S, q: Queue[(K, V)]): ZIO[TamerDBConfig, TamerError, S] =
+    (for {
+      log   <- logTask
+      cfg   <- ConfigDb.queryConfig
+      tnx   <- ZIO.service[Transactor[Task]]
+      query <- UIO(setup.queryBuilder.query(state))
+      _     <- log.debug(s"running ${query.sql} with params derived from $state")
+      start <- UIO(Instant.now())
+      values <-
+        query
+          .streamWithChunkSize(cfg.fetchChunkSize)
+          .chunks
+          .transact(tnx)
+          .map(ChunkWithMetadata(_))
+          .evalTap(c => q.offerAll(toIterable(c.chunk).map(v => setup.keyExtract(v) -> v)))
+          .flatMap(c => Stream.chunk(c.chunk).map(ValueWithMetadata(_, c.pulledAt)))
+          .compile
+          .toList
+      pulledTimeOrNow = values.headOption.map(_.pulledAt).getOrElse(Instant.now())
+      newState <- setup.stateFoldM(state)(
+        QueryResult(
+          ResultMetadata(
+            Duration.between(start, pulledTimeOrNow).toMillis
+          ),
+          values.map(_.value)
+        )
+      )
+    } yield newState).mapError(e => TamerError(e.getLocalizedMessage, e))
+
+  final def fetchWithTimeSegment[K <: Product: Encoder: Decoder: SchemaFor, V <: Product with Timestamped: Ordering: Encoder: Decoder: SchemaFor](
+      queryBuilder: TimeSegment => Query0[V]
+  )(earliest: Instant, tumblingStep: Duration, keyExtract: V => K): ZIO[ZEnv, TamerError, Unit] = {
+    val setup = mkSetupWithTimeSegment[K, V](queryBuilder)(earliest, tumblingStep, keyExtract)
+    fetch(setup)
+  }
+
+  final def fetch[
+      K <: Product: Encoder: Decoder: SchemaFor,
+      V <: Product: Encoder: Decoder: SchemaFor,
+      S <: Product with HashableState: Encoder: Decoder: SchemaFor
+  ](
+      setup: Setup[K, V, S]
+  ): ZIO[ZEnv, TamerError, Unit] = {
+    val kafkaLayer: Layer[TamerError, Kafka]                                      = Config.live >>> Kafka.live
+    val transactorLayer: Layer[TamerError, DbTransactor]                          = (Blocking.live ++ ConfigDb.live) >>> db.hikariLayer
+    val queryConfigLayer: Layer[TamerError, DbConfig with QueryConfig]            = ConfigDb.live
+    val defaultLayer: Layer[TamerError, DbTransactor with Kafka with QueryConfig] = transactorLayer ++ queryConfigLayer ++ kafkaLayer
+    tamer.kafka.runLoop(setup)(iteration(setup)).provideCustomLayer(defaultLayer)
+  }
+
+  val hikariLayer: ZLayer[Blocking with DbConfig, TamerError, DbTransactor] = ZLayer.fromManaged {
+    for {
+      cfg               <- ConfigDb.dbConfig.toManaged_
+      connectEC         <- ZIO.descriptor.map(_.executor.asEC).toManaged_
+      blockingEC        <- blocking.blocking(ZIO.descriptor.map(_.executor.asEC)).toManaged_
+      managedTransactor <- mkTransactor(cfg, connectEC, blockingEC)
+    } yield managedTransactor
+  }
+
+  def mkTransactor(db: ConfigDb.Db, connectEC: ExecutionContext, transactEC: ExecutionContext): Managed[TamerError, HikariTransactor[Task]] =
```

Comment:
do we want to expose `mkTransactor` as part of the public api ?

Reply:
I’m assuming not